test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,926 @@
+
+In the second step, we look at "effective" `@since` tags in the mainline sources, from `src.zip`
+ (if the test run doesn't have it, we throw a `jtreg.SkippedException`)
+    - We only check the specific MODULE whose name was passed as an argument in the test.
+      In that module, we look for unqualified exports and test those packages.
+    - The `@since` checker verifies that for every API element, the real since value and
+      the effective since value are the same, and reports an error if they are not.
+
+Important note : We only check code written since JDK 9 as the releases used to determine the expected value
+                 of @since tags are taken from the historical data built into javac which only goes back that far
+
+note on rules for Real and effective `@since:
+
+Real since value of an API element is computed as the oldest release in which the given API element was introduced.
+That is:
+- for modules, packages, classes and interfaces, the release in which the element with the given qualified name was introduced
+- for constructors, the release in which the constructor with the given VM descriptor was introduced
+- for methods and fields, the release in which the given method or field with the given VM descriptor became a member
+  of its enclosing class or interface, whether direct or inherited
+
+Effective since value of an API element is computed as follows:
+- if the given element has a @since tag in its javadoc, it is used
+- in all other cases, return the effective since value of the enclosing element
+
+
+Special Handling for preview method, as per JEP 12:
+- When an element is still marked as preview, the `@since` should be the first JDK release where the element was added.
+- If the element is no longer marked as preview, the `@since` should be the first JDK release where it was no longer preview.
+
+note on legacy preview: Until JDK 14, the preview APIs were not marked in any machine-understandable way.
+                        It was deprecated, and had a comment in the javadoc.
+                        and the use of `@PreviewFeature` only became standard in JDK 17.
+                        So the checker has an explicit knowledge of these preview elements.
+
+note: The `<unique-Element-ID>` for methods looks like
+      `method: <erased-return-descriptor> <binary-name-of-enclosing-class>.<method-name>(<ParameterDescriptor>)`.
+it is somewhat inspired from the VM Method Descriptors. But we use the erased return so that methods
+that were later generified remain the same.
+*/
+
+public class SinceChecker {
+    private final Map<String, Set<String>> LEGACY_PREVIEW_METHODS = new HashMap<>();
+    private final Map<String, IntroducedIn> classDictionary = new HashMap<>();
+    private final JavaCompiler tool;
+    private int errorCount = 0;
+
+    // packages to skip during the test
+    private static final Set<String> EXCLUDE_LIST = Set.of(
+            "java.lang.classfile"
+    );
```

Comment:
This should not really be here, in the general code for a SinceChecker for all modules.

It would be better if this info was provided in command-line options, given when the test is run from the module-specific test (i.e. `CheckSince_javaBase.java`). 

This means you will have to write some additional simple option decoding in the `main` method, starting line 21.